File: chess/src/main/scala/Status.scala
package lila.chess

sealed abstract class Status(val id: Int) extends Ordered[Status] {

  def compare(other: Status) = id compare other.id
}

case object Created extends Status(10)
case object Started extends Status(20)
case object Aborted extends Status(25) // from this point the game is finished
case object Mate extends Status(30)
case object Resign extends Status(31)
case object Stalemate extends Status(32)
case object Timeout extends Status(33) // when player leaves the game
case object Draw extends Status(34)
case object Outoftime extends Status(35) // clock flag
case object Cheat extends Status(36)

object Status {

  val all = List(Created, Started, Aborted, Mate, Resign, Stalemate, Timeout, Draw, Outoftime, Cheat)

  val byId = all map { v ⇒ (v.id, v) } toMap

  def apply(id: Int): Option[Status] = byId get id
}

File: first_repo/ui/game/src/status.ts
import { GameData } from './interfaces';

// https://github.com/ornicar/scalachess/blob/master/src/main/scala/Status.scala

export const ids = {
  created: 10,
  started: 20,
  aborted: 25,
  mate: 30,
  resign: 31,
  stalemate: 32,
  timeout: 33,
  draw: 34,
  outoftime: 35,
  cheat: 36,
  noStart: 37,
  variantEnd: 60,
};

export function started(data: GameData): boolean {
  return data.game.status.id >= ids.started;
}

export function finished(data: GameData): boolean {
  return data.game.status.id >= ids.mate;
}

export function aborted(data: GameData): boolean {
  return data.game.status.id === ids.aborted;
}

export function playing(data: GameData): boolean {
  return started(data) && !finished(data) && !aborted(data);
}
